test/tools/d_do_test.d

```diff
@@ -962,11 +966,94 @@ int runBashTest(string input_dir, string test_name)
     version(Windows)
     {
         auto process = spawnShell(format("bash %s %s %s",
-            buildPath(scriptDir, "tools", "sh_do_test.sh"), input_dir, test_name));
+            buildPath(dmdTestDir, "tools", "sh_do_test.sh"), input_dir, test_name));
     }
     else
     {
-        auto process = spawnProcess([scriptDir.buildPath("tools", "sh_do_test.sh"), input_dir, test_name]);
+        auto process = spawnProcess([dmdTestDir.buildPath("tools", "sh_do_test.sh"), input_dir, test_name]);
     }
     return process.wait();
 }
+
+int runDShellTest(string input_dir, string test_name, const ref EnvData envData,
+    string output_dir, string output_file)
+{
+    const testScriptDir = buildPath(dmdTestDir, input_dir);
+    const testScriptPath = buildPath(testScriptDir, test_name ~ ".d");
+    const dshellPath = buildPath(dmdTestDir, "tools", "dshell.d");
+    const dshellSource = readText(dshellPath);
+
+    writefln(" ... %s/%s.d", input_dir, test_name);
+    if (exists(output_file))
+        remove(output_file);
+
+    // create the "dshellglue" module to glue dshell and the 'test' together
+    const glueSourceDir = buildPath(output_dir, test_name ~ ".dshellglue");
+    mkdirRecurse(glueSourceDir);
+    {
+        auto glueSource = File(buildPath(glueSourceDir, "dshellglue.d"), "w");
+        glueSource.writeln(`module dshellglue;`);
+        glueSource.writefln(`enum TEST_DIR  = "%s";`, input_dir);
+        glueSource.writefln(`enum TEST_NAME = "%s";`, test_name);
+    }
```

Comment:
Move in a separate function instead of creating a sub-scope?

Reply:
Meh, that means creatinga new function and passing at least 3 variables to it for 4 lines of code...

Reply:
I believe doing that would make the code-flow harder to understand. This may be a 100-line function but it has no loops/jumps, the code just flows from top to bottom. Just a few if statements with a max depth of 2 and a few blocks to separate some variables into sub-scopes. Having long functions is fine, it is more about what the function is doing I think.

By splitting the code into multiple functions now you have to start checking/verifying/understanding what the state is in each method. You can't assume that a function that should have run before unless you go through all the code and verify it is always called before you. It provides less guarantees as now if you require something to run before you, you now need to verify that thing is always called before you. All that being said, if the function doesn't require this then it's not really an issue, but having the code be in it's own function stills adds more strain to the reader since just having it in a function means it can be called at any time, where with a long function, the reader knows implicitly when that code is going to be executed in relation to everything else.

Reply:
FWIW I was thinking about sth. like:

```d
struct DShellTester {
   void run() {
     initialize();
     createDShellModules();
     compileTests();
     runTests();
     cleanupArtifacts();
   }
}
```

where the reader immediately gets an high-level overview of what's happening and can then dive into the function he is actually interested in. The fact that you already use scopes for grouping shows that visual grouping is necessary ;-)

Reply:
Ok here it is.  For me this is MUCH harder to follow.  It seems unnecessarily complex and convoluted.  The single function was much simpler and you could immediately tell what was going to be executed and in what order without having to understand multiple functions.  But maybe this one is easier for you to follow?